File: src/leetcode/P139_WordBreak.java
package leetcode;

 import java.util.HashMap;
import java.util.List;

/*
    Given a string s and a dictionary of words dict, 
    determine if s can be segmented into a space-separated 
    sequence of one or more dictionary words.

	For example, given
	s = "leetcode",
	dict = ["leet", "code"].
	
	Return true because "leetcode" can be segmented as "leet code".
 */

/**
 * @author      zxwtry
 * @email       zxwtry@qq.com
 * @project     OJ
 * @package     leetcode
 * @file        P139_WordBreak.java
 * @date        2017年5月29日 上午10:20:42
 * @details     Solution1: AC 18ms 41.08%
 */
public class P139_WordBreak {
    static class Solution1 {
        public boolean wordBreak(String s, List<String> wd) {
            int sn = s == null ? 0 : s.length();
            HashMap<String, Boolean> nm = new HashMap<>(sn);
            for (String w : wd) nm.put(w, true);
            return search(s, 0, sn, nm);
        }
        private boolean search(String s, int si, int sn, HashMap<String, Boolean> nm) {
            if (si == sn) return true;
            String res = s.substring(si, sn);
            Boolean resVal = nm.get(res);
            if (resVal != null && (!resVal)) return false;
            boolean ans = false;
            for (int i = si+1; (!ans) && i <= sn; i ++) {
                String n = s.substring(si, i);
                Boolean nVal = nm.get(n);
                if (nVal != null && nVal)
                    ans |= search(s, i, sn, nm);
            }
            nm.put(res, ans);
            return ans;
        }
    }
}

File: python/P139_WordBreak.py
#coding=utf-8

'''
    url: leetcode.com/problems/word-break
    @author:     zxwtry
    @email:      zxwtry@qq.com
    @date:       2017年5月29日
    @details:    Solution:  49ms 84.54%
'''

class Solution(object):
    def wordBreak(self, s, d):
        """
        :type s: str
        :type d: List[str]
        :rtype: bool
        """
        sn = 0 if s == None else len(s)
        dp = [False] * (sn + 1)
        dp[0] = True
        for i in range(sn):
            if not dp[i]: continue
            for w in d:
                l = len(w)
                if (i+l <= sn and (not dp[i+l]) and s[i :i+l] == w):
                    dp[i+l] = True
        return dp[sn]

